core/src/main/java/io/opentelemetry/android/export/BufferedDelegatingExporter.kt

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.android.export
+
+import io.opentelemetry.sdk.common.CompletableResultCode
+import java.util.concurrent.atomic.AtomicBoolean
+
+/**
+ * An in-memory buffer delegating signal exporter that buffers signal in memory until a delegate is set.
+ * Once a delegate is set, the buffered signals are exported to the delegate.
+ *
+ * The buffer size is set to 5,000 by default. If the buffer is full, the exporter will drop new signals.
+ */
+internal abstract class BufferedDelegatingExporter<T, D>(private val bufferedSignals: Int = 5_000) {
+    @Volatile
+    private var delegate: D? = null
+    private val buffer = arrayListOf<T>()
+    private val lock = Any()
+    private var isShutDown = AtomicBoolean(false)
+
+    /**
+     * Sets the delegate for this exporter and flushes the buffer to the delegate.
+     *
+     * If the delegate has already been set, an [IllegalStateException] will be thrown.
+     * If this exporter has been shut down, the delegate will be shut down immediately.
+     *
+     * @param delegate the delegate to set
+     *
+     *  @throws IllegalStateException if a delegate has already been set
+     */
+    fun setDelegate(delegate: D) {
+        synchronized(lock) {
+            check(this.delegate == null) { "Exporter delegate has already been set." }
+
+            flushToDelegate(delegate)
+
+            this.delegate = delegate
+
+            if (isShutDown.get()) {
+                shutdownDelegate(delegate)
+            }
+        }
+    }
+
+    /**
+     * Buffers the given data if the delegate has not been set, otherwise exports the data to the delegate.
+     *
+     * @param data the data to buffer or export
+     */
+    protected fun bufferOrDelegate(data: Collection<T>): CompletableResultCode =
+        withDelegateOrNull {
+            if (it != null) {
+                exportToDelegate(it, data)
+            } else {
+                val amountToTake = bufferedSignals - buffer.size
+                buffer.addAll(data.take(amountToTake))
+                CompletableResultCode.ofSuccess()
```

Comment:
Yeah I suppose, but there might be _a lot_ of warnings. But I you probably want to know. How do you prefer we log this? I would avoid logging using `Log` since logging in production is frowned upon.

Reply:
@breedx-splk What would you like to do here? I think logging is a good idea but not in production. Is there an existing solution for this already?

Reply:
I still think that if/when this happens it's going to be challenging to see without ANY signs...so maybe doing a one-shot log is better than nothing. Kind-of a stupid amount of extra work, but something like this:

```
private val haveLoggedErrorOnce = AtomicBoolean(false)
...
if(haveLoggedErrorOnce.compareAndSet(false, true)) {
    Log.w(OTEL_RUM_LOG_TAG,"Telemetry buffer was filled before export delegate set...")
    Log.w(OTEL_RUM_LOG_TAG,"This has resulted in a loss of telemetry data!")
}
```

Reply:
I suppose these days it doesn't matter too much since logcat filtering is much better than it was. I wouldn't really mind logging every time honestly.

Also, you could imagine a small logging framework that otel ships with that you can hook into. I'd love to use open telemetry isefl to log this but if logcat is used there is no way to hook in. But this sounds like an issue for a later date.

Reply:
Thanks yeah I'm fine with doing the simpler thing and just logging it every time, but I thought you were reluctant. 🙃 
Anyway, yeah, probably worth mentioning that we've had a log related issue here https://github.com/open-telemetry/opentelemetry-android/issues/142 for a year, looking for contributors who want to take that on.

Reply:
If we're gonna log we might as well log haha